metals/src/main/scala/scala/meta/internal/metals/ExcludedPackagesHandler.scala

```diff
@@ -0,0 +1,43 @@
+package scala.meta.internal.metals
+
+/**
+ * This class handles packages that should be excluded from indexing causing
+ * them to not be recommended for completions, symbol searches, and code actions.
+ *
+ * @param userConfig access to user input for excluded packages
+ */
+class ExcludedPackagesHandler(userConfig: () => UserConfiguration) {
+  val defaultExclusions: List[String] = List(
+    "META-INF/", "images/", "toolbarButtonGraphics/", "jdk/", "sun/", "javax/",
+    "oracle/", "java/awt/desktop/", "org/jcp/", "org/omg/", "org/graalvm/",
+    "com/oracle/", "com/sun/", "com/apple/", "apple/"
+  )
+
+  private def handleUserExclusions(pkgs: List[String]): List[String] = {
+    pkgs.foldRight(defaultExclusions)((pkg, acc) => {
+      val replaced = pkg.replace(".", "/")
+      val formatted =
+        if (replaced.endsWith("/")) replaced
+        else replaced + "/"
+
+      if (formatted.startsWith("--")) {
+        acc.filterNot(_.equals(formatted.drop(2)))
+      } else {
+        formatted :: acc
+      }
+    })
+  }
+
+  /**
+   * Should the given package be excluded from indexing
+   *
+   * @param pkg package to check against excluded list
+   */
+  def isExcludedPackage(pkg: String): Boolean = {
+    val packagesToExclude: List[String] = userConfig().excludedPackages match {
+      case Some(pkgs) => handleUserExclusions(pkgs)
```

Comment:
I think this method-call here will need to be cached to avoid performance regressions. This is a fairly hot code path that's triggered for all completions and workspace/symbol.

Reply:
That was my original thought as well, but I've put logging in here to test, and this only ever seems to be hit on when we index. It's appears to not be hit on when triggering completions, symbol searches etc.

Reply:
You’re right. It’s probably fine as it is. I would still try to optimize the foldLeft, I’m generally paranoid with using immutable collections on hot paths like indexing.

Reply:
👍 totally. I was able to change this around a bit and cache it because when testing it on a large code base it was getting called hundreds of times. So that should be taken care of now.